### ISSUE TITLE
How to solve maximum number of connected components reached in 3D ?

### ISSUE THREAD

**Issue by marbre2** [Image-Analysis-Hub/cellpose-appose#91]
I use latest version of fiji (java21) with Fiji-cellpose plugin on a set of image.
In this issue, I have a small volume 7 slices stack where I run Cellpose models (cyto3 and cpsam) in order to compare which performs best in the task of dtecting small ablong objects (typ diameter 5-10pix).

I encounter the following issue that I identified while changing the LUT of the labels image. By default, the output is a 16b image.
In my case, the cellpose model tend to oversegment all the small objects and seems to reach the 65535 limit.

I tried to convert afterward to a 32bit image in order to solve the 3D connection of the object but might not have found the right solution so far.

**Comment by StRigaud** [comment]
Hi @marbre2  :)

In the current state of the plugin I do not think it is possible as the return type is defined as Integers hence 16-bit for Fiji.
We need to see if we can change the code to allow 32-bit output

**Comment by gletort** [comment]
Thansk @marbre2 ! Yes, we have to see if we can add a 32-bit output option. The best might be to have a checkbox in the parameters asking for a 32-bit output, because apparently there is no way to know before what cellpose will return 16-bit or 32-bit. @tinevez, you had a look at the data format, do you think it makes sense to add an option to ask for a 32-bit output  and use that to create the shared memory for appose ?

**Comment by tinevez** [comment]
Hi all.
Yes I looked. Indeed, Cellpose returns a 16-bit label image **or** a 32-bit, and you cannot know what will be the output type before you run `model.eval()`. But since we are calling it from Python we can act there, before returning to Fiji.

We can add another button to specify the return type. And it could be many more things beyond `32-bit` float numbers. 
I would be in favor of instead condensing the labels, and using the `n-colors` lib of Kevin Cutler. 
OR, we could put all the options in the `imglib2-cellpose` and make opinionated choices of what we present to the user in Fiji-Cellpose.

**Comment by gletort** [comment]
The issue is not for the colors, so I don't see why using the n-colors ? And it would make labels not unique. But yes we could have several options in the imglib2-cellpose, in some applications the labels may not have to be unique.

**Comment by tinevez** [comment]
And probably the issue won't be limited to Cellpose, so we probably need to abstract the solution.
So to adapt the label image output, in Python, before returning to Java, we have the following options that the user could configure:

- Cast on 
    - `uint16`
    - `uint32`
    - `uint64`
    - or `float32`
(IJ1 does not support `uint32` and `uint64`)

- Randomize labels: yes or no (see https://github.com/imglib/imglib2-cellpose/blob/main/src/main/resources/net/imglib2/cellpose/cp_utils.py#L72-L86)

- Reassign labels: yes or no (à la `ncolor` https://github.com/kevinjohncutler/ncolor)
'yes' generate a label image with at most N different colors, so we could cast the output on a `uint8` to save RAM

- Enforce label unicity: yes or no (see https://github.com/imglib/imglib2-cellpose/blob/main/src/main/resources/net/imglib2/cellpose/cp_utils.py#L54-L69)

We can add options for these 4 parameters in https://github.com/imglib/imglib2-cellpose/blob/main/src/main/java/net/imglib2/cellpose/CellposeParameters.java
Questions: 

- In what order to deal with them in the Python code? 
- What should we show to a IJ1 plugin user?

**Comment by gletort** [comment]
I'm not sure that I understand what you mean. We have the enforce label unicity option and the randomize options in imglib2-cellpose now, as parameters. For this 32-bit option, from what I understood, it's already possible in imglib2-cellpose, so we just have to add this option in Fiji-Cellpose as a parameter in the GUI to select a 32-bit  output, right ? And then in the fiji-cellpose code call properly imglib2-cellpose 3 or 4 with the outputType and  CellposeOutput casted with UnsignedIntType or UnsignedShortType.
